server/src/main/java/org/elasticsearch/common/filesystem/PosixFileSystemNatives.java

```diff
@@ -0,0 +1,138 @@
+import com.sun.jna.Native;
+import com.sun.jna.NativeLibrary;
+import com.sun.jna.Platform;
+
+import com.sun.jna.Pointer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Constants;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.OptionalLong;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * {@link FileSystemNatives.Provider} implementation for Linux x86-64bits
+ */
+final class PosixFileSystemNatives implements FileSystemNatives.Provider {
+
+    private static final Logger logger = LogManager.getLogger(PosixFileSystemNatives.class);
+
+    private static final PosixFileSystemNatives INSTANCE = new PosixFileSystemNatives();
+
+    /** st_blocks field indicates the number of blocks allocated to the file, 512-byte units **/
+    private static final long ST_BLOCKS_UNIT = 512L;
+
+    /**
+     * Version of the `struct stat' data structure.
+     *
+     * To allow the `struct stat' structure bits to vary without changing shared library major version number, the `stat' function is often
+     * an inline wrapper around `xstat' which takes a leading version-number argument designating the data structure and bits used.
+     *
+     * In glibc this version is defined in bits/stat.h (or bits/struct_stat.h in glibc 2.33, or bits/xstatver.h in more recent versions).
+     *
+     * For x86-64 the _STAT_VER used is:
+     *  # define _STAT_VER_LINUX    1
+     *  # define _STAT_VER _STAT_VER_LINUX
+     *
+     * For other architectures the _STAT_VER used is:
+     *  # define _STAT_VER_LINUX    0
+     *  # define _STAT_VER _STAT_VER_LINUX
+     **/
+    private static int loadStatVersion() {
+        return "aarch64".equalsIgnoreCase(Constants.OS_ARCH) ? 0 : 1;
+    }
+
+    private final Stat64Library lib;
+
+    private PosixFileSystemNatives() {
+        assert Constants.JRE_IS_64BIT : Constants.OS_ARCH;
+        Stat64Library statFunction;
+        try {
+            var libc = NativeLibrary.getInstance(Platform.C_LIBRARY_NAME);
+            libc.getFunction("stat64");
+            // getfunction didn't fail, so the symbol is available
+            statFunction = Native.load(Platform.C_LIBRARY_NAME, Stat64Library.class);
+        } catch (UnsatisfiedLinkError e) {
+            var xstat = Native.load(Platform.C_LIBRARY_NAME, XStatLibrary.class);
+            var version =loadStatVersion();
+            statFunction = (path, stats) -> xstat.__xstat(version, path, stats);
+        }
+        lib = statFunction;
+        logger.debug("C library loaded");
+    }
+
+    static PosixFileSystemNatives getInstance() {
+        return INSTANCE;
+    }
+
+    public interface Stat64Library extends Library {
+        int stat64(String path, Pointer stats) throws LastErrorException;
+    }
+
+    public interface XStatLibrary extends Library {
+        int __xstat(int version, String path, Pointer stats) throws LastErrorException;
```

Comment:
Again, more for my curiosity, I see this changed to Pointer from Stat. This makes sense, Stat is not a JNA Structure anymore, but I don't understand why the difference with fcntl above.

Reply:
I changed Stat to no longer be a Structure, so JNA won't convert it to a pointer. See the call site, it passes stats.memory.